File: rln/ffi_nim_examples/multi_message_id.nim
include "common"

proc createMessageIds(ids: array[maxOut, uint32]): Vec_Fr =
  result = ffi_vec_fr_new(csize_t(maxOut))
  for id in ids:
    let tmp = ffi_uint_to_fr(id)
    ffi_vec_fr_push(addr result, tmp)
    ffi_fr_free(tmp)

proc createMultiWitness(member: Member,
    merkleProof: ptr MerkleProof, messageIds: ptr Vec_Fr,
    selectorUsed: var array[maxOut, bool], x: ptr Fr,
    externalNullifier: ptr Fr): WitnessResult =
  var selectorVec = Vec_bool(dataPtr: addr selectorUsed[0],
      len: csize_t(maxOut), cap: csize_t(maxOut))
  ffi_rln_witness_input_new_multi(member.identitySecret,
      member.userMessageLimit, messageIds, merkleProof, x, externalNullifier,
      addr selectorVec)

proc main() =
  var rlnInstance = initRLN(true)
  if rlnInstance.isNil:
    return

  var member = createMember()

  let merkleProof = registerMember(rlnInstance, member.rateCommitment)
  if merkleProof.isNil:
    return

  let externalNullifier = computeExternalNullifier()

  echo "\nHashing first signal"
  var signal1: array[32, uint8] = [1'u8, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0, 0, 0, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  let x1 = hashSignal(signal1)
  printFr("x1", x1)

  echo "\nCreating first message ids and selector used"
  echo "  - using 2 out of " & $maxOut & " slots"
  var messageIds1 = createMessageIds([0'u32, 1, 0, 0])
  var selectorUsed1 = [true, true, false, false]
  printVecFr("message ids", addr messageIds1)

  echo "\nCreating first RLN witness"
  let witness1Result = createMultiWitness(member, merkleProof,
      addr messageIds1, selectorUsed1, x1, externalNullifier)
  if witness1Result.ok.isNil:
    stderr.writeLine("First witness creation error: " & asString(
        witness1Result.err))
    ffi_c_string_free(witness1Result.err)
    return
  var witness1 = witness1Result.ok
  echo "  - first RLN witness created successfully"

  echo "\nGenerating first RLN proof"
  let rlnProof1Result = ffi_rln_generate_proof(rlnInstance, witness1)
  if rlnProof1Result.ok.isNil:
    stderr.writeLine("Proof generation error: " & asString(rlnProof1Result.err))
    ffi_c_string_free(rlnProof1Result.err)
    return
  var rlnProof1 = rlnProof1Result.ok
  echo "  - proof generated successfully"

  echo "\nGetting first RLN proof values"
  var proofValues1 = ffi_rln_proof_get_values(rlnProof1)
  let ys1Result = ffi_rln_proof_values_get_ys(proofValues1)
  if ys1Result.err.dataPtr != nil:
    stderr.writeLine("Get ys error: " & asString(ys1Result.err))
    ffi_c_string_free(ys1Result.err)
    return
  var ys1 = ys1Result.ok
  printVecFr("ys", addr ys1)
  ffi_vec_fr_free(ys1)
  let nullifiers1Result = ffi_rln_proof_values_get_nullifiers(proofValues1)
  if nullifiers1Result.err.dataPtr != nil:
    stderr.writeLine("Get nullifiers error: " & asString(nullifiers1Result.err))
    ffi_c_string_free(nullifiers1Result.err)
    return
  var nullifiers1 = nullifiers1Result.ok
  printVecFr("nullifiers", addr nullifiers1)
  ffi_vec_fr_free(nullifiers1)
  let selectorUsedResult = ffi_rln_proof_values_get_selector_used(
      proofValues1)
  if selectorUsedResult.err.dataPtr != nil:
    stderr.writeLine("Get selector used error: " & asString(
        selectorUsedResult.err))
    ffi_c_string_free(selectorUsedResult.err)
    return
  let proofSelectorUsed = selectorUsedResult.ok
  var selectorStr = ""
  for i in 0 ..< int(proofSelectorUsed.len):
    if i > 0:
      selectorStr.add(", ")
    let slot = cast[ptr UncheckedArray[bool]](proofSelectorUsed.dataPtr)[i]
    selectorStr.add(if slot: "true" else: "false")
  echo "  - selector used = [" & selectorStr & "]"
  ffi_vec_bool_free(proofSelectorUsed)
  let proofValues1Root = ffi_rln_proof_values_get_root(proofValues1)
  printFr("root", proofValues1Root)
  ffi_fr_free(proofValues1Root)
  let proofValues1X = ffi_rln_proof_values_get_x(proofValues1)
  printFr("x", proofValues1X)
  ffi_fr_free(proofValues1X)
  let proofValues1ExternalNullifier =
    ffi_rln_proof_values_get_external_nullifier(proofValues1)
  printFr("external nullifier", proofValues1ExternalNullifier)
  ffi_fr_free(proofValues1ExternalNullifier)

  echo "\nVerifying first proof"
  let verify1Result = verifyStatefulProof(rlnInstance, rlnProof1, x1)
  if verify1Result.err.dataPtr != nil:
    stderr.writeLine("Proof verification error: " & asString(verify1Result.err))
    ffi_c_string_free(verify1Result.err)
    return
  if verify1Result.ok:
    echo "  - first proof verified successfully"
  else:
    echo "First proof verification failed"
    return

  echo "\nSimulating double-signaling attack (same epoch, different message)"

  echo "\nHashing second signal"
  var signal2: array[32, uint8] = [11'u8, 12, 13, 14, 15, 16, 17, 18, 19, 20, 0,
      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
  let x2 = hashSignal(signal2)
  printFr("x2", x2)

  echo "\nCreating second message ids and selector used"
  echo "  - using 2 out of " & $maxOut & " slots"
  echo "  - duplicated slot id 1"
  var messageIds2 = createMessageIds([1'u32, 0, 3, 0])
  var selectorUsed2 = [true, false, true, false]
  printVecFr("message ids", addr messageIds2)

  echo "\nCreating second RLN witness"
  let witness2Result = createMultiWitness(member, merkleProof,
      addr messageIds2, selectorUsed2, x2, externalNullifier)
  if witness2Result.ok.isNil:
    stderr.writeLine("Second witness creation error: " & asString(
        witness2Result.err))
    ffi_c_string_free(witness2Result.err)
    return
  var witness2 = witness2Result.ok
  echo "  - second RLN witness created successfully"

  echo "\nGenerating second RLN proof"
  let rlnProof2Result = ffi_rln_generate_proof(rlnInstance, witness2)
  if rlnProof2Result.ok.isNil:
    stderr.writeLine("Second proof generation error: " & asString(
        rlnProof2Result.err))
    ffi_c_string_free(rlnProof2Result.err)
    return
  var rlnProof2 = rlnProof2Result.ok
  var proofValues2 = ffi_rln_proof_get_values(rlnProof2)
  echo "  - second proof generated successfully"

  echo "\nVerifying second proof"
  let verify2Result = verifyStatefulProof(rlnInstance, rlnProof2, x2)
  if verify2Result.err.dataPtr != nil:
    stderr.writeLine("Proof verification error: " & asString(verify2Result.err))
    ffi_c_string_free(verify2Result.err)
    return
  if verify2Result.ok:
    echo "  - second proof verified successfully"

    echo "\nRecovering identity secret"
    let recoverResult = ffi_rln_recover_id_secret(proofValues1,
        proofValues2)
    if recoverResult.ok.isNil:
      stderr.writeLine("Identity recovery error: " & asString(
          recoverResult.err))
      ffi_c_string_free(recoverResult.err)
      return
    let recoveredSecret = recoverResult.ok
    if ffi_secret_fr_eq(recoveredSecret, member.identitySecret):
      let recoveredDebug = ffi_secret_fr_debug(recoveredSecret)
      echo "  - recovered secret = " & asString(recoveredDebug) &
          " matches the original identity secret"
      ffi_c_string_free(recoveredDebug)
    ffi_secret_fr_free(recoveredSecret)
  else:
    echo "Second proof verification failed"

  ffi_rln_proof_values_free(proofValues2)
  ffi_rln_proof_free(rlnProof2)
  ffi_rln_witness_input_free(witness2)
  ffi_vec_fr_free(messageIds2)
  ffi_fr_free(x2)
  ffi_rln_proof_values_free(proofValues1)
  ffi_rln_proof_free(rlnProof1)
  ffi_rln_witness_input_free(witness1)
  ffi_vec_fr_free(messageIds1)
  ffi_fr_free(x1)
  ffi_fr_free(externalNullifier)
  ffi_rln_merkle_proof_free(merkleProof)
  memberFree(member)
  ffi_rln_free(rlnInstance)

main()

File: rln/ffi_c_examples/multi_message_id.c
#include <stdio.h>
#include <stdlib.h>

#include "common.c"

static Vec_Fr create_message_ids(const unsigned int ids[MAX_OUT])
{
    Vec_Fr message_ids = ffi_vec_fr_new(MAX_OUT);
    for (size_t i = 0; i < MAX_OUT; i++)
    {
        Fr *tmp = ffi_uint_to_fr(ids[i]);
        ffi_vec_fr_push(&message_ids, tmp);
        ffi_fr_free(tmp);
    }
    return message_ids;
}

static WitnessResult
create_multi_witness(const Member *member, const MerkleProof *merkle_proof,
                     const Vec_Fr *message_ids, bool selector_used[MAX_OUT], const Fr *x,
                     const Fr *external_nullifier)
{
    return ffi_rln_witness_input_new_multi(member->identity_secret,
                                           member->user_message_limit, message_ids,
                                           merkle_proof, x, external_nullifier,
                                           &(Vec_bool){selector_used, MAX_OUT, MAX_OUT});
}

int main(void)
{
    RLN *rln_instance = init_rln(true);
    if (!rln_instance)
    {
        return EXIT_FAILURE;
    }

    Member member;
    create_member(&member);

    MerkleProof *merkle_proof = register_member(rln_instance, member.rate_commitment);
    if (!merkle_proof)
    {
        return EXIT_FAILURE;
    }

    Fr *external_nullifier = compute_external_nullifier();

    printf("\nHashing first signal\n");
    uint8_t signal1[32] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0, 0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    Fr *x1 = hash_signal(signal1);
    print_fr("x1", x1);

    printf("\nCreating first message ids and selector used\n");
    printf("  - using 2 out of %d slots\n", MAX_OUT);
    const unsigned int ids1[MAX_OUT] = {0, 1, 0, 0};
    Vec_Fr message_ids1 = create_message_ids(ids1);
    bool selector_used1[MAX_OUT] = {true, true, false, false};
    print_vec_fr("message ids", &message_ids1);

    printf("\nCreating first RLN witness\n");
    WitnessResult witness1_result = create_multi_witness(
        &member, merkle_proof, &message_ids1, selector_used1, x1, external_nullifier);
    if (!witness1_result.ok)
    {
        fprintf(stderr, "First witness creation error: %.*s\n",
                (int)witness1_result.err.len, (char *)witness1_result.err.ptr);
        ffi_c_string_free(witness1_result.err);
        return EXIT_FAILURE;
    }
    Witness *witness1 = witness1_result.ok;
    printf("  - first RLN witness created successfully\n");

    printf("\nGenerating first RLN proof\n");
    ProofResult rln_proof1_result =
        ffi_rln_generate_proof(rln_instance, witness1);
    if (!rln_proof1_result.ok)
    {
        fprintf(stderr, "Proof generation error: %.*s\n",
                (int)rln_proof1_result.err.len, (char *)rln_proof1_result.err.ptr);
        ffi_c_string_free(rln_proof1_result.err);
        return EXIT_FAILURE;
    }
    Proof *rln_proof1 = rln_proof1_result.ok;
    printf("  - proof generated successfully\n");

    printf("\nGetting first RLN proof values\n");
    ProofValues *proof_values1 = ffi_rln_proof_get_values(rln_proof1);
    VecFrResult ys1_result = ffi_rln_proof_values_get_ys(proof_values1);
    if (ys1_result.err.ptr)
    {
        fprintf(stderr, "Get ys error: %.*s\n",
                (int)ys1_result.err.len, (char *)ys1_result.err.ptr);
        ffi_c_string_free(ys1_result.err);
        return EXIT_FAILURE;
    }
    print_vec_fr("ys", &ys1_result.ok);
    ffi_vec_fr_free(ys1_result.ok);
    VecFrResult nullifiers1_result =
        ffi_rln_proof_values_get_nullifiers(proof_values1);
    if (nullifiers1_result.err.ptr)
    {
        fprintf(stderr, "Get nullifiers error: %.*s\n",
                (int)nullifiers1_result.err.len, (char *)nullifiers1_result.err.ptr);
        ffi_c_string_free(nullifiers1_result.err);
        return EXIT_FAILURE;
    }
    print_vec_fr("nullifiers", &nullifiers1_result.ok);
    ffi_vec_fr_free(nullifiers1_result.ok);
    VecBoolResult proof_selector_used_result =
        ffi_rln_proof_values_get_selector_used(proof_values1);
    if (proof_selector_used_result.err.ptr)
    {
        fprintf(stderr, "Get selector used error: %.*s\n",
                (int)proof_selector_used_result.err.len,
                (char *)proof_selector_used_result.err.ptr);
        ffi_c_string_free(proof_selector_used_result.err);
        return EXIT_FAILURE;
    }
    Vec_bool proof_selector_used = proof_selector_used_result.ok;
    printf("  - selector used = [");
    for (size_t i = 0; i < proof_selector_used.len; i++)
    {
        printf("%s%s", i ? ", " : "", proof_selector_used.ptr[i] ? "true" : "false");
    }
    printf("]\n");
    ffi_vec_bool_free(proof_selector_used);
    Fr *proof_values1_root = ffi_rln_proof_values_get_root(proof_values1);
    print_fr("root", proof_values1_root);
    ffi_fr_free(proof_values1_root);
    Fr *proof_values1_x = ffi_rln_proof_values_get_x(proof_values1);
    print_fr("x", proof_values1_x);
    ffi_fr_free(proof_values1_x);
    Fr *proof_values1_external_nullifier =
        ffi_rln_proof_values_get_external_nullifier(proof_values1);
    print_fr("external nullifier", proof_values1_external_nullifier);
    ffi_fr_free(proof_values1_external_nullifier);

    printf("\nVerifying first proof\n");
    CBoolResult verify1_result = verify_stateful_proof(rln_instance, rln_proof1, x1);
    if (verify1_result.err.ptr)
    {
        fprintf(stderr, "Proof verification error: %.*s\n",
                (int)verify1_result.err.len, (char *)verify1_result.err.ptr);
        ffi_c_string_free(verify1_result.err);
        return EXIT_FAILURE;
    }
    if (verify1_result.ok)
    {
        printf("  - first proof verified successfully\n");
    }
    else
    {
        printf("First proof verification failed\n");
        return EXIT_FAILURE;
    }

    printf("\nSimulating double-signaling attack (same epoch, different message)\n");

    printf("\nHashing second signal\n");
    uint8_t signal2[32] = {11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 0, 0, 0, 0, 0, 0,
                           0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    Fr *x2 = hash_signal(signal2);
    print_fr("x2", x2);

    printf("\nCreating second message ids and selector used\n");
    printf("  - using 2 out of %d slots\n", MAX_OUT);
    printf("  - duplicated slot id 1\n");
    const unsigned int ids2[MAX_OUT] = {1, 0, 3, 0};
    Vec_Fr message_ids2 = create_message_ids(ids2);
    bool selector_used2[MAX_OUT] = {true, false, true, false};
    print_vec_fr("message ids", &message_ids2);

    printf("\nCreating second RLN witness\n");
    WitnessResult witness2_result = create_multi_witness(
        &member, merkle_proof, &message_ids2, selector_used2, x2, external_nullifier);
    if (!witness2_result.ok)
    {
        fprintf(stderr, "Second witness creation error: %.*s\n",
                (int)witness2_result.err.len, (char *)witness2_result.err.ptr);
        ffi_c_string_free(witness2_result.err);
        return EXIT_FAILURE;
    }
    Witness *witness2 = witness2_result.ok;
    printf("  - second RLN witness created successfully\n");

    printf("\nGenerating second RLN proof\n");
    ProofResult rln_proof2_result =
        ffi_rln_generate_proof(rln_instance, witness2);
    if (!rln_proof2_result.ok)
    {
        fprintf(stderr, "Second proof generation error: %.*s\n",
                (int)rln_proof2_result.err.len, (char *)rln_proof2_result.err.ptr);
        ffi_c_string_free(rln_proof2_result.err);
        return EXIT_FAILURE;
    }
    Proof *rln_proof2 = rln_proof2_result.ok;
    ProofValues *proof_values2 = ffi_rln_proof_get_values(rln_proof2);
    printf("  - second proof generated successfully\n");

    printf("\nVerifying second proof\n");
    CBoolResult verify2_result = verify_stateful_proof(rln_instance, rln_proof2, x2);
    if (verify2_result.err.ptr)
    {
        fprintf(stderr, "Proof verification error: %.*s\n",
                (int)verify2_result.err.len, (char *)verify2_result.err.ptr);
        ffi_c_string_free(verify2_result.err);
        return EXIT_FAILURE;
    }
    if (verify2_result.ok)
    {
        printf("  - second proof verified successfully\n");

        printf("\nRecovering identity secret\n");
        SecretFrResult recover_result =
            ffi_rln_recover_id_secret(proof_values1, proof_values2);
        if (!recover_result.ok)
        {
            fprintf(stderr, "Identity recovery error: %.*s\n",
                    (int)recover_result.err.len, (char *)recover_result.err.ptr);
            ffi_c_string_free(recover_result.err);
            return EXIT_FAILURE;
        }
        SecretFr *recovered_secret = recover_result.ok;
        if (ffi_secret_fr_eq(recovered_secret, member.identity_secret))
        {
            Vec_uint8 recovered_debug = ffi_secret_fr_debug(recovered_secret);
            printf("  - recovered secret = %.*s matches the original identity secret\n",
                   (int)recovered_debug.len, (char *)recovered_debug.ptr);
            ffi_c_string_free(recovered_debug);
        }
        ffi_secret_fr_free(recovered_secret);
    }
    else
    {
        printf("Second proof verification failed\n");
    }

    ffi_rln_proof_values_free(proof_values2);
    ffi_rln_proof_free(rln_proof2);
    ffi_rln_witness_input_free(witness2);
    ffi_vec_fr_free(message_ids2);
    ffi_fr_free(x2);
    ffi_rln_proof_values_free(proof_values1);
    ffi_rln_proof_free(rln_proof1);
    ffi_rln_witness_input_free(witness1);
    ffi_vec_fr_free(message_ids1);
    ffi_fr_free(x1);
    ffi_fr_free(external_nullifier);
    ffi_rln_merkle_proof_free(merkle_proof);
    member_free(&member);
    ffi_rln_free(rln_instance);
    return EXIT_SUCCESS;
}
